content/tutorial-nlp-from-scratch.ipynb

```diff
@@ -0,0 +1,1458 @@
+    "\n",
+    "        # store values needed for backward propagation in cache\n",
+    "        cache = (next_hidden_state, next_cell_state,\n",
+    "                 prev_hidden_state, prev_cell_state,\n",
+    "                 ft, it, cmt, ot, X_t)\n",
+    "        caches['lstm_values'].append(cache)\n",
+    "\n",
+    "        # Update hidden state and cell state for next time step\n",
+    "        prev_hidden_state = next_hidden_state\n",
+    "        prev_cell_state = next_cell_state\n",
+    "\n",
+    "    # Pass through a fully connected layer to perform binary classification\n",
+    "    z2 = (np.dot(parameters['W2'], next_hidden_state)\n",
+    "          + parameters['b2'])\n",
+    "    a2 = sigmoid(z2)\n",
+    "    cache = (a2, parameters['W2'])\n",
+    "    caches['fc_values'].append(cache)\n",
+    "\n",
+    "    return caches"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cc168c0c",
+   "metadata": {},
+   "source": [
+    "### Backpropagation\n",
+    "\n",
+    "After each forward pass through the network, you will implement the `backpropagation through time` algorithm to accumulate gradients of each parameter over the time steps. Backpropagation through a LSTM is not as straightforward as through other common Deep Learning architectures, due to the special way its underlying layers interact. Nonetheless, the approach is largely the same; identifying dependencies and applying the chain rule."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "fa9d7000",
+   "metadata": {},
+   "source": [
+    "Lets start with defining a function to initialise gradients of each parameter as arrays made up of zeros with same dimensions as the corresponding parameter"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 26,
+   "id": "04e541af",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "# Initialise the gradients\n",
+    "def initialise_grads(parameters):\n",
+    "    grads = {}\n",
+    "    for param in parameters.keys():\n",
+    "        grads[f'd{param}'] = np.zeros((parameters[param].shape))\n",
+    "    return grads"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f2794880",
+   "metadata": {},
+   "source": [
+    "Now you'll define a function to calculate the gradients of each intermediate value in the neural network with respect to the loss and accumulate those gradients over the entire sequence. To understand how the gradients are calculated at each step in greater depth, you are suggested to follow this helpful [blog](https://christinakouridi.blog/2019/06/19/backpropagation-lstm/) by Christina Kouridi"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "id": "fba4c540",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "def backprop(y, caches, hidden_dim, input_dim, time_steps, parameters):\n",
+    "    # Retrieve output and corresponding weights of fully connected layer\n",
```

Comment:
May be this can be broken down into different functions according to the gates present in LSTM. For example, first a forget gate is present which takes current word in the sequence (x_t) and the hidden state of previous cell (h_t_prev). A function can be added,
```python
def forget_gate(x_t, h_t_prev):
    # do stuff
   return forget
```
There are lots of `x`, `+`, sigma, `tanh` which have their own role in the LSTM model, so writing functions with meaningful names for each of them will be great. These can be then called at appropriately places inside `backprop` and forward pass function (don't know the name of the function implemented for this). Let me know if you want to discuss this further.

"Step by Step LSTM Walk Through" section in https://colah.github.io/posts/2015-08-Understanding-LSTMs/ can help in modularising this thing. Cheers.

Reply:
Hi @czgdp1807, apologies for the late response. I really liked your suggestions and have incorporated the suggested changes in the latest commit along with the addition of a small diagram that gives an overview of the gate mechanisms happening inside a memory block of the LSTM. Let me know what you think!

Reply:
Looks good. It's easier to understand now.